### PR TITLE
CATROID-870: Sorting variables and lists

### PR DESCRIPTION
**Pull request by khushal1707**
https://jira.catrob.at/browse/CATROID-870
Sorting and Undo-Sorting of DataList from overflow menu.
(Issue with this implementation: While list is sorted, on adding new variable\list Recyclerview didn't update and on deletion of variable\list app crashes) 

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -2075,5 +2075,8 @@ needs read and write access to it. You can always change permissions through you
     <string name="serif_font" translatable="false">serif</string>
     <string name="edit">Edit</string>
     <string name="data_value">Value</string>
+    <string name="undo_sort">Undo Sort</string>
+    <string name="sort_data">Sort</string>
+    <string name="sort">Sort</string>
```
Comment (wslany) [inline catroid/src/main/res/values/strings.xml:2080]:
```suggestion
    <string name="undo_sort">Undo sort</string>
    <string name="sort">Sort</string>
```

Why sort_data and sort, two times the same?

**Review by wslany — CHANGES_REQUESTED** [review]
Some comment, and please rewrite FormulaEditorSortDataListTest.java into a kotlin test. Thanks!

File: catroid/src/main/res/values/strings.xml (hunk 1 above)
Comment (khushal1707) [reply catroid/src/main/res/values/strings.xml:2080]: Working on the changes, also correcting above mentioned error, thanks for notifying

**Comment by khushal1707** [comment]
FormulaEditorSortDataListTest.java changed to kotlin, comments added and remove double initialization, please review it again, Thanks!

**Comment by wslany** [comment]
Sorry for the late reply, I was busy and didn't check the ticket because it is not yet in the normal workflow. 

First, there are 15 lint warning. Can you please fix them? Details see 
https://jenkins.catrob.at/job/Catroid/job/PR-3972/8/analysis/#issuesContent

**Comment by khushal1707** [comment]
> First, there are 15 lint warning. Can you please fix them? Details see
> https://jenkins.catrob.at/job/Catroid/job/PR-3972/8/analysis/#issuesContent
> 
> Should be very easy to fix.

Done!
Actually the issue of deletion after sorting is not fixed yet, the problem that I understood is that after sorting deleting item in list is not seems unsupported and throwing UnsupportedOperationException.
It would be great if I get some guidance on the same by fellow developers of catrobat, here (on github) or on slack if possible.
Thanks!

File: catroid/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -2075,5 +2075,7 @@ needs read and write access to it. You can always change permissions through you
     <string name="serif_font" translatable="false">serif</string>
     <string name="edit">Edit</string>
     <string name="data_value">Value</string>
+    <string name="undo_sort">Undo Sort</string>
```
Comment (wslany) [inline catroid/src/main/res/values/strings.xml:2078]:
```suggestion
    <string name="undo_sort">Undo sort</string>
```

**Review by wslany — COMMENTED** [review]
Thanks! Could you check whether the sorting order also works with Hindi (or Gujarati or what your language is)?

Regarding the UnsupportedOperationException, I will see what I can do and put you in contact with someone who can help.

**Comment by khushal1707** [comment]
> Thanks! Could you check whether the sorting order also works with Hindi (or Gujarati or what your language is)?
> 
> Regarding the UnsupportedOperationException, I will see what I can do and put you in contact with someone who can help.

Sure, I will check this in Hindi also

File: catroid/src/main/res/values/strings.xml (hunk 2 above)
Comment (khushal1707) [reply catroid/src/main/res/values/strings.xml]: Done

**Comment by wslany** [comment]
Thanks. Regarding the UnsupportedOperationException, does this only happen with your PR? Which part of the code, and could you paste a stack trace?

**Comment by khushal1707** [comment]
Yes, the issue is associated with this PR only and can be reproduced by sorting data in data list and deleting variable.

**Stack trace:**
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.catrobat.catroid, PID: 6238
    java.lang.UnsupportedOperationException
        at java.util.AbstractList.remove(AbstractList.java:161)
        at java.util.AbstractList$Itr.remove(AbstractList.java:374)
        at java.util.AbstractCollection.remove(AbstractCollection.java:293)
        at org.catrobat.catroid.ui.recyclerview.adapter.RVAdapter.remove(RVAdapter.java:139)
        at org.catrobat.catroid.ui.recyclerview.adapter.DataListAdapter.remove(DataListAdapter.java:365)
        at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment.deleteItems(DataListFragment.kt:310)
        at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment.access$deleteItems(DataListFragment.kt:58)
        at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment$showDeleteAlert$1.onClick(DataListFragment.kt:300)
        at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6692)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
   E/BaseExceptionHandler: uncaughtException: 
   java.lang.UnsupportedOperationException
        at java.util.AbstractList.remove(AbstractList.java:161)
        at java.util.AbstractList$Itr.remove(AbstractList.java:374)
        at java.util.AbstractCollection.remove(AbstractCollection.java:293)
        at org.catrobat.catroid.ui.recyclerview.adapter.RVAdapter.remove(RVAdapter.java:139)
        at org.catrobat.catroid.ui.recyclerview.adapter.DataListAdapter.remove(DataListAdapter.java:365)
        at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment.deleteItems(DataListFragment.kt:310)
        at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment.access$deleteItems(DataListFragment.kt:58)
        at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment$showDeleteAlert$1.onClick(DataListFragment.kt:300)
        at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6692)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```
And the first code reference link in the above error (RVAdapter.java:139):
```java
public boolean remove(T item) {
		if (items.remove(item)) {
			notifyItemRemoved(items.indexOf(item));
			return true;
		}
		return false;
	}
```

**Comment by wslany** [comment]
OK, I just contacted you via your gmail account, please check, let's follow up there.

Great that it works fine with Hindi.

**Comment by khushal1707** [comment]
> OK, I just contacted you via your gmail account, please check, let's follow up there.
> 
> Great that it works fine with Hindi.

Sure and thanks you very much for the invite!

**Comment by ndrnour** [comment]
> Yes, the issue is associated with this PR only and can be reproduced by sorting data in data list and deleting variable.
> Also this implementation is working fine on Hindi.
> 
> **Stack trace:**
> 
> ```
> E/AndroidRuntime: FATAL EXCEPTION: main
>     Process: org.catrobat.catroid, PID: 6238
>     java.lang.UnsupportedOperationException
>         at java.util.AbstractList.remove(AbstractList.java:161)
>         at java.util.AbstractList$Itr.remove(AbstractList.java:374)
>         at java.util.AbstractCollection.remove(AbstractCollection.java:293)
>         at org.catrobat.catroid.ui.recyclerview.adapter.RVAdapter.remove(RVAdapter.java:139)
>         at org.catrobat.catroid.ui.recyclerview.adapter.DataListAdapter.remove(DataListAdapter.java:365)
>         at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment.deleteItems(DataListFragment.kt:310)
>         at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment.access$deleteItems(DataListFragment.kt:58)
>         at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment$showDeleteAlert$1.onClick(DataListFragment.kt:300)
>         at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
>         at android.os.Handler.dispatchMessage(Handler.java:106)
>         at android.os.Looper.loop(Looper.java:193)
>         at android.app.ActivityThread.main(ActivityThread.java:6692)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
>    E/BaseExceptionHandler: uncaughtException: 
>    java.lang.UnsupportedOperationException
>         at java.util.AbstractList.remove(AbstractList.java:161)
>         at java.util.AbstractList$Itr.remove(AbstractList.java:374)
>         at java.util.AbstractCollection.remove(AbstractCollection.java:293)
>         at org.catrobat.catroid.ui.recyclerview.adapter.RVAdapter.remove(RVAdapter.java:139)
>         at org.catrobat.catroid.ui.recyclerview.adapter.DataListAdapter.remove(DataListAdapter.java:365)
>         at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment.deleteItems(DataListFragment.kt:310)
>         at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment.access$deleteItems(DataListFragment.kt:58)
>         at org.catrobat.catroid.ui.recyclerview.fragment.DataListFragment$showDeleteAlert$1.onClick(DataListFragment.kt:300)
>         at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
>         at android.os.Handler.dispatchMessage(Handler.java:106)
>         at android.os.Looper.loop(Looper.java:193)
>         at android.app.ActivityThread.main(ActivityThread.java:6692)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
> ```
> 
> And the first code reference link in the above error (RVAdapter.java:139):
> 
> ```java
> public boolean remove(T item) {
> 		if (items.remove(item)) {
> 			notifyItemRemoved(items.indexOf(item));
> 			return true;
> 		}
> 		return false;
> 	}
> ```

the reason for getting UnsupportedOperationException that you are trying to modify an unmodifiable List! 
basically `sortedWith(comparator: Comparator<in T>)` extension function return an unmodifiable List.
how to fix that?
very simple, just create an array list of the current list, for example:
`userDefinedBrickInputs = userDefinedBrickInputs.sortedWith(Comparator { item1:            UserDefinedBrickInput, item2: UserDefinedBrickInput ->                item1.name.compareTo(item2.name)            })`
`userDefinedBrickInputs = ArrayList(userDefinedBrickInputs) // new extra line which fix your problem`

**Comment by khushal1707** [comment]
> the reason for getting UnsupportedOperationException that you are trying to modify an unmodifiable List!
> basically `sortedWith(comparator: Comparator<in T>)` extension function return an unmodifiable List.
> how to fix that?
> very simple, just create an array list of the current list, for example:
> `userDefinedBrickInputs = userDefinedBrickInputs.sortedWith(Comparator { item1: UserDefinedBrickInput, item2: UserDefinedBrickInput -> item1.name.compareTo(item2.name) })`
> `userDefinedBrickInputs = ArrayList(userDefinedBrickInputs) // new extra line which fix your problem`

@ndrnour This definitely resolves the error but doesn't solves the issue, if we delete variable from the new list it will still remain in original list and after updating recyclerview it'll again show up in the list.
Try this: create a list -> sort it -> delete a variable -> Un-sort it, 
deleted variable will again show up.
As `sortedWith` method is creating immutable list that why I tried using `Collections.sort` and here is PR for that https://github.com/Catrobat/Catroid/pull/4174, please have a look and if there is any other way to resolve above mentioned issue please mention it, thanks for looking into it :)

**Pull request closed without merging**